web/packages/teleport/src/lib/useMfa.ts

```diff
@@ -77,27 +99,67 @@ export function useMfa(emitterSender: EventEmitterMfaSender): MfaState {
       });
   }
 
-  const onChallenge = useCallback(challengeJson => {
-    const { webauthnPublicKey, ssoChallenge, totpChallenge } =
-      makeMfaAuthenticateChallenge(challengeJson);
-
-    setState(prevState => ({
-      ...prevState,
-      ssoChallenge,
-      webauthnPublicKey,
-      totpChallenge,
-    }));
-  }, []);
+  const waitForSsoChallengeResponse = useCallback(
+    async (
+      ssoChallenge: SSOChallenge,
+      abortSignal: AbortSignal
+    ): Promise<void> => {
+      const channel = new BroadcastChannel(ssoChallenge.channelId);
+
+      try {
+        const event = await waitForMessage(channel, abortSignal);
+        emitterSender.sendChallengeResponse({
+          sso_response: {
+            requestId: ssoChallenge.requestId,
+            token: event.data.mfaToken,
+          },
+        });
+        clearChallenges();
+      } catch (error) {
+        if (error.name !== 'AbortError') {
+          throw error;
+        }
+      } finally {
+        channel.close();
+      }
+    },
+    [emitterSender]
+  );
 
   useEffect(() => {
-    if (emitterSender) {
-      emitterSender.on(TermEvent.WEBAUTHN_CHALLENGE, onChallenge);
-
-      return () => {
-        emitterSender.removeListener(TermEvent.WEBAUTHN_CHALLENGE, onChallenge);
-      };
-    }
-  }, [emitterSender, onChallenge]);
+    let ssoChallengeAbortController: AbortController | undefined;
+    const chanllengeHandler = (challengeJson: string) => {
+      const { webauthnPublicKey, ssoChallenge, totpChallenge } =
+        makeMfaAuthenticateChallenge(challengeJson);
+
+      setState(prevState => ({
+        ...prevState,
+        addMfaToScpUrls: true,
+        ssoChallenge,
+        webauthnPublicKey,
+        totpChallenge,
+      }));
+
+      if (ssoChallenge) {
+        ssoChallengeAbortController?.abort();
+        ssoChallengeAbortController = new AbortController();
+        void waitForSsoChallengeResponse(
+          ssoChallenge,
+          ssoChallengeAbortController.signal
+        );
+      }
```

Comment:
nit: It seems to me that `onSsoAuthenticate` is a better place for this logic. We don't really need to setup this listener until the user actually clicks SSO and wants to go through with it. Could we move all this `abortController` logic and waiting there?

Reply:
We need the abortController logic here because we need a way to abort if the component unmounts (in the cleanup function of this useEffect).